torch/lib/c10d/CUDAUtils.cpp

```diff
@@ -7,10 +7,16 @@ namespace c10d {
 CUDAEvent CUDAEvent::create(unsigned int flags) {
   CUDAEvent event;
   C10D_CUDA_CHECK(cudaEventCreateWithFlags(&event.event_, flags));
+  int current_device;
+  C10D_CUDA_CHECK(cudaGetDevice(&current_device));
+  event.setDevice(current_device);
   return event;
 }
 
 CUDAEvent::~CUDAEvent() {
+  // cudaEventDestroy must run on the same device of the event,
+  // otherwise it creates a context on default device as well.
+  at::DeviceGuard guard(device_);
```

Comment:
Can you move this into the if block? Then the cost is going to be almost 0 when we did a std::swap to assign an event. With this change it will still execute a cudaSetDevice twice.